### ISSUE TITLE
Suggestions on RapidCRC

### ISSUE THREAD

**Issue by kuwwoon** [OV2/RapidCRC-Unicode#78]
Hi, I like the tool!

I have some issues using the tool as it is:

1. If I hit ESC it quits, unexpected as for me, without warning, but I have lots of calculated hashes in there and spent a lot of time generating them.
2. When I mark a string with a file and a hash it remains selected and I am not able to exit selection mode, so I am not able to save all the hashes except the selected one. Crtl+A doesn't work in the main window during selection. The only way in this case is to go to the first file and then Shift+END to select all the files. Maybe the ESC button is worth to set with the selecting mode exit function if a user entered it. Leave Exit and Close Button to do this exit job, maybe with a YES/NO warning.
3. I would set to ask where to save a hash-list file first and then launch the calculation if the list is empty. I can leave the job going on but know when several hours pass it will be ready and saved.

Thanks! 

Andriy, forensic examiner

**Comment by OV2** [comment]
Exiting on esc is standard dialog behavior, maybe I can add an option.

You can deselect by clicking on a location in the list without files. I'll try to add a "select all".

Switching the dialog-calc order might not fit in the current design, I'll take a look.

**Comment by kuwwoon** [comment]
Thanks, for prompt response. I am considering the tool as default for mere law enforcement detectives. They need a tool to verify they copied files from one media to another correctly. I tried a lot of tools, but this one is the best on simplicity.
I will follow the conversation to learn if you managed to add some of my suggestions. Many thanks!

**Comment by kuwwoon** [comment]
Also I haven't managed to use the tool against the file with a hash list. Even the feature Put CRC to filename also not available for definite .md5, .sha256 and so on files.
Colleagues need to write down some checksum for the file with a hash list. Thanks!

**Comment by OV2** [comment]
I don't completely get your last message. What do you mean with "use the tool against the file with a hash list"?
Do you want to calculate the hash of an sfv/md5 file itself? That currently only works if it is not the first file dropped/selected in the open dialog.

**Comment by kuwwoon** [comment]
You have got what I mean correctly. I've noticed that behavior as well. But it is not obvious. Maybe I get an idea what to do to solve the issue. Thanks.

**Comment by kuwwoon** [comment]
Maybe let all the function from content menu like Create ... or Put .... work with all file like target for hashing, even if it is the only md5 or sha... file.

**Comment by OV2** [comment]
Currently you can use that to operate on all files specified in the sfv file, so this would remove functionality. But maybe I can add an entry to force open all files as regular files in the popup menu when you rightclick-drag (or hold ctrl) files into the list. And also into the open dialog dropdown menu.

**Comment by kuwwoon** [comment]
> 
> 
> Currently you can use that to operate on all files specified in the sfv file, so this would remove functionality. But maybe I can add an entry to force open all files as regular files in the popup menu when you rightclick-drag (or hold ctrl) files into the list. And also into the open dialog dropdown menu.

It would be useful because I am considering we need an additional tool to calculate a hash for a file list. Having so great tool it should be able to do it somehow as well.

**Comment by kuwwoon** [comment]
Hi. Could you please look at the behavior of the functions inside the tool. Provided I have sent to the tool a bunch of files without hashing them. I've got only file list. Then I decided to create a hashlist, let say SHA256. I launch the button and get the warning that I need to calculate hashsums for the set of files. I hit ok and after hashing process I get nothing. I need to hit the button again. Personally, I expect to get a windows with a choice where to save a hashlist. Thanks!

**Comment by kuwwoon** [comment]
Could you please describe the rules on how to launch a tool in different modes it can provide. I have PCs without Admin rights and the tool works perfect on them. I could create some links we need with predefined modes like those you provide at a contextual menu. Thanks!

**Comment by kuwwoon** [comment]
> Exiting on esc is standard dialog behavior, maybe I can add an option.
> 
> You can deselect by clicking on a location in the list without files. I'll try to add a "select all".
> 
> Switching the dialog-calc order might not fit in the current design, I'll take a look.

That would be great to put a warning window before exit. Thanks!

**Comment by OV2** [comment]
> I launch the button and get the warning that I need to calculate hashsums for the set of files. I hit ok and after hashing process I get nothing. I need to hit the button again.
That was the behavior of the original rcrc. I'll take a look.

> Could you please describe the rules on how to launch a tool in different modes it can provide.
What modes do you mean? You want to pass in files via the cmd line and use something like "open all hash files"?

**Comment by kuwwoon** [comment]
> > I launch the button and get the warning that I need to calculate hashsums for the set of files. I hit ok and after hashing process I get nothing. I need to hit the button again.
> 
> That was the behavior of the original rcrc. I'll take a look.
> 
> > Could you please describe the rules on how to launch a tool in different modes it can provide.
> 
> What modes do you mean? You want to pass in files via the cmd line and use something like "open all hash files"?

Yes, it seems you got me right. I guess all the available modes via menu have some options behind to run rapidcrc.exe, like rapidcrc.exe /sha256.

**Comment by OV2** [comment]
Unfortunately that is not possible at the moment, rcrc is not really built for running it from the command line.

**Comment by kuwwoon** [comment]
Hi again! I like the feature Open all hash files, but at the Status line I can see only results. If I get corrupt file or missing one and that file is not from the last hash file, I am not able to determine what file are in question. If you be kind to consider adding files to a general list. Thanks!

**Comment by OV2** [comment]
That sounds like you're using open all hash files without having enabled job queueing. If you enable that, multiple jobs can be displayed in the list.

**Comment by kuwwoon** [comment]
> That sounds like you're using open all hash files without having enabled job queueing. If you enable that, multiple jobs can be displayed in the list.

Hi! You are right. It works as you wrote, thanks! But what the meaning of a function  - New window in explorer?

I would like to ask you to consider to add a save the check results botton/function for each hash file checked. It can first contain file information and a column with check result, and then calculated hashes. Thanks!

**Comment by kuwwoon** [comment]
Hi again! I was wondering why do you put the portable files to RAR archive? Anyway Chrome downloads this files via Save link as ... only. Many users do not have advanced archiver, and default Windows archiver supports only zip. I would recommend to put the portable files to zip archive instead to let every user extract portable version at any rate. Thanks!

**Comment by OV2** [comment]
> Hi! You are right. It works as you wrote, thanks! But what the meaning of a function - New window in explorer?

When you enable job queueing, doubleclicking a sfv file in explorer (or using the shell extension) will usually add a new job to the queue of an existing rapidcrc window. This option will force rcrc to open a new window.

> I would like to ask you to consider to add a Save check results botton/function for each hash file processed/checked. [...]

I'm not sure I understand what would be different with this than a regular sfv / md5 file?

**Comment by OV2** [comment]
> Hi again! I was wondering why do you put the portable files to RAR archive? 

It is simply a format I've always used and that compressed well. Since it is supported by 7zip, I've never seen it as a problem. There is also the installer that should be usable by everyone.

**Comment by kuwwoon** [comment]
> > I would like to ask you to consider to add a Save check results botton/function for each hash file processed/checked. [...]
> 
> I'm not sure I understand what would be different with this than a regular sfv / md5 file?

You are right. I haven't guessed that if I go to a hash button and hit the appropriate hash name I can save the new results as they are. Thank you for the tip!

Talking about the option =Include file information comments=, I would suggest putting also a value from the Info column to the comments like File Corrupt or File Ok or File not found. It is useful to document that at the moment of a check files were checked against another hash file with definite results. Secondly, that is worth mentioning is timestamps. If I save a hash file with comments against files on the NTFS file system I get timestamps according to the time zone at my PC, that can be wrong. And that is the problem because such an approach can misrepresent the actual facts. I would suggest putting the additional information to the time like 21:26.23 2021-01-11 (NTFS;UTC) or (FAT; local) and put UTC time instead of local for NTFS files (or to have a switch at the settings menu to be able to select what timestamp approach should be utilized).
Anyway, thank you for your time improving the tool!

**Comment by kuwwoon** [comment]
> 
> 
> > Hi again! I was wondering why do you put the portable files to RAR archive?
> 
> It is simply a format I've always used and that compressed well. Since it is supported by 7zip, I've never seen it as a problem. There is also the installer that should be usable by everyone.

You know, I use often a RAR format as well. It is a commercial one and it gets the work done in all cases. I use it also with its recovery feature. And you are right that 7-Zip is a great free tool, but not in all cases people have 7-Zip installed on their work PCs, which are usually restricted to install new software even like 7-Zip. I have tested RAR vs Zip and RAR shows an advantage (384/404 Kb), but the overall size is so small that can be ignored. Sure, I do not insist, but it would be nice to have portable versions in Zip. Thanks!

**Comment by kuwwoon** [comment]
Great tool, amazing! Best wishes to OV2!
I use the tool for working with digital evidence, it helps a lot!

**Comment by kuwwoon** [comment]
OV2, thank you for the new release. I will test all the improvements soon. But even now I see many suggestions were implemented, even zip-archive is in place on your web-page now. Your tool helps greatly in digital forensic job. I teach people how easy it is to use the tool in every day activities. Great!

**Comment by sergeevabc** [comment]
I don’t mean to sound rude, but one can't throw words out of a song, so here is the thing: RapidCRC’s interface is so f---ed up and non-intuitive _for years_. For example, why there is no ability to select calculated pairs filenames-hashes and copy them to clipboard with a space in between to paste, let’s say, in a messenger window? Why there is no basic Ctrl+C, damn?

**Comment by kuwwoon** [comment]
> I don’t mean to sound rude, but one can't throw words out of a song, so here is the thing: RapidCRC’s interface is so f---ed up and non-intuitive _for years_. For example, why there is no ability to select calculated pairs filenames-hashes and copy them to clipboard with a space in between to paste, let’s say, in a messenger window? Why there is no basic Ctrl+C, damn?

You were indeed rude providing such a comment! The tool is freeware, so you didn't pay for it. You can use it as it is or refuse using the tool. But your feature request is appropriate, I cannot but agree. OV2, it would be nice to have the feature implemented. Thanks!

**Comment by OV2** [comment]
> why there is no ability to select calculated pairs filenames-hashes and copy them to clipboard

It should be obvious why it is not there (I did not need it and nobody requested it). But you're not really asking that, you want the feature.

In what format should it be copied to the clipboard (there could be multiple files selected with multiple hashes calculated)? A header row, then one file on each line, with tab separating columns (good for pasting in excel for example)? Or filename on one line, then the calculated hashes below (maybe one per line), repeating for all files? Or filename, then all hashes simply separated by space (but then how do you know which hash is which - the name in front of each hash)?

**Comment by kuwwoon** [comment]
I have noticed that the tool verifies the integrity of the just copied files using their images in RAM. The verification speed points on that. If I remove target drive where files has been copied I get realistic speed rate. Please, have a look on that. Desired behavior - to verify reading from the very drive even if RAM still holds the just copied data. Thank you!

**Comment by OV2** [comment]
RCRC is not really meant to verify that file write operations from memory work correctly. That's something your copy tool should do if you really need it.
I might look into adding an option to do unbuffered reads, but that is more work than simply specifying a flag when opening the file.

**Comment by kuwwoon** [comment]
> RCRC is not really meant to verify that file write operations from memory work correctly. That's something your copy tool should do if you really need it.
> I might look into adding an option to do unbuffered reads, but that is more work than simply specifying a flag when opening the file.

As mentioned several months earlier, we use the tool as convenient forensic tool to verify integrity of just  copied files from a source drive to a target drive. The results of verification is used then as a proof of integrity of copied data. That allows us to represent copies in court as admissible evidence if a source drive gets broken or lost. We would appreciate to have a function of unbuffered reading embedded as an option to the tool. Thank you!

**Comment by kuwwoon** [comment]
@OV2 have you decided wether you are going to add unbuffered read option to the tool?

**Comment by OV2** [comment]
I will add the option, but I can't tell you when I'll find the time to work on RCRC again.

**Comment by OV2** [comment]
I've uploaded a new version that contains an option for unbuffered reads.

**Comment by kuwwoon** [comment]
Again, thank you for the tool! Hundreds of civil servants use it for the common good.
I often use it as portable version. But I miss an option how to load a folder with files or other folder and files from the menu. Could you please consider adding the function "Open folder". Thanks!

**Comment by OV2** [comment]
You can either open a folder via the shell extension, or drag it onto to window from explorer. If you need to do it via the open dialog this is only possible by going into the folder and then selecting everything in it and pressing open.

**Comment by gppdee2p** [comment]
Hi! Dragging it onto to window from explorer is the solution. Thanks!

пн, 1 серп. 2022, 23:12 користувач OV2 ***@***.***> пише:
